crates/ruff_db/src/program.rs

```diff
@@ -0,0 +1,85 @@
+// TODO: Fix clippy warnings in Salsa macros
+#![allow(clippy::needless_lifetimes, clippy::clone_on_copy)]
+
+use crate::{system::SystemPathBuf, Db};
+
+#[salsa::input(singleton)]
+pub struct Program {
```

Comment:
It's not really clear to me why we should have both `Program` and `Workspace` as separate things. I realize that currently `Program` is a true singleton and `Workspace` is not, but the only reason for that is that we might want to hold onto the old Workspace when constructing a new one, in case we fail to build the new one, and it seems like this same rationale applies to `Program` just as much -- it is built from user settings that we might fail to parse. So what is the benefit of having `Program` and `Workspace` be separate, and what is the logic that defines what is owned by `Program` and what is owned by `Workspace`?

Reply:
I explained one motivation in the summary:

> The new Program ingredient (singleton) mainly generalizes the existing module resolution settings and moves them into ruff_db. The motivation is that these settings aren't just important for module resolution. They're fundamental to the entire analysis. Two analyses using different program settings aren't compatible and should be cached (persistent) separately.

That generally means that `Workspace` holds on to significantly more information than `Program`. It also solves a technical problem. Workspace must hold on to the `Configuration`/`Settings,` but the `Configuration` depends on the `LinterConfiguration` and `FormatterConfiguration` structs that are defined in `ruff_linter` and `ruff_python_formatter`. `ruff_db` can't depend on `ruff_linter`, because `ruff_linter` depends on `ruff_db`. The alternative is not to have a `Program` ingredient and instead keep per-crate specific inputs as we used to have with the module-resolver. But I prefer not to do that because input ingredients require manual creation (and updating).

What should be on `Program`? Just the values that define the *identity* of a `Program`. Another way to think about it is that two projects can't be "checked" together if they require a different value for any of the `Program` fields.